### PR TITLE
*: initial commit of openvswitch_exporter

### PR DESCRIPTION
**Pull request by mdlayher**
Let's do this thing.

### CODE REVIEW

**Comment by mdlayher** [comment]
```
mlayher@[redacted]:~$ curl -s http://localhost:9310/metrics | grep openvswitch
# HELP openvswitch_datapath_info Metadata about a given datapath.
# TYPE openvswitch_datapath_info gauge
openvswitch_datapath_info{datapath="ovs-system",index="6"} 1
# HELP openvswitch_datapath_megaflow_stats_mask_hits_total Number of megaflow masks used for flow lookups.
# TYPE openvswitch_datapath_megaflow_stats_mask_hits_total counter
openvswitch_datapath_megaflow_stats_mask_hits_total{datapath="ovs-system"} 1.1698109634e+12
# HELP openvswitch_datapath_megaflow_stats_masks_total Number of megaflow masks present.
# TYPE openvswitch_datapath_megaflow_stats_masks_total counter
openvswitch_datapath_megaflow_stats_masks_total{datapath="ovs-system"} 82
# HELP openvswitch_datapath_stats_flows Number of flows present.
# TYPE openvswitch_datapath_stats_flows gauge
openvswitch_datapath_stats_flows{datapath="ovs-system"} 4838
# HELP openvswitch_datapath_stats_hits_total Number of flow table matches.
# TYPE openvswitch_datapath_stats_hits_total counter
openvswitch_datapath_stats_hits_total{datapath="ovs-system"} 1.98010649735e+11
# HELP openvswitch_datapath_stats_lost_total Number of flow table misses not sent to userspace.
# TYPE openvswitch_datapath_stats_lost_total counter
openvswitch_datapath_stats_lost_total{datapath="ovs-system"} 47418
# HELP openvswitch_datapath_stats_misses_total Number of flow table misses.
# TYPE openvswitch_datapath_stats_misses_total counter
openvswitch_datapath_stats_misses_total{datapath="ovs-system"} 1.558781205e+09
```

File: internal/ovsexporter/datapath.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+	const (
+		subsystem = "datapath"
+	)
+
+	var (
+		labels = []string{"datapath"}
+	)
+
+	return &datapathCollector{
+		Info: prometheus.NewDesc(
```
Comment (mdlayher) [inline internal/ovsexporter/datapath.go]: Now that I removed the "features" field (didn't seem relevant to us), perhaps I should just remove this metric entirely.  It appears that the datapath enforces unique names, and the interface index doesn't add much value on its own.

File: internal/ovsexporter/ovsexporter.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	cs []prometheus.Collector
+}
+
+var _ prometheus.Collector = &collector{}
+
+// New creates a new Prometheus collector which collects metrics using the
+// input Open vSwitch generic netlink client.
+func New(c *ovsnl.Client) prometheus.Collector {
+	return &collector{
+		cs: []prometheus.Collector{
```
Comment (TimSimmons) [inline internal/ovsexporter/ovsexporter.go:42]: Why a slice here, might there be multiple collectors from a single client in the future?

File: internal/ovsexporter/datapath_test.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+				`openvswitch_datapath_megaflow_stats_mask_hits_total{datapath="ovs-system"} 5`,
+				`openvswitch_datapath_megaflow_stats_masks_total{datapath="ovs-system"} 6`,
+				`openvswitch_datapath_stats_flows{datapath="ovs-system"} 4`,
+				`openvswitch_datapath_stats_hits_total{datapath="ovs-system"} 1`,
+				`openvswitch_datapath_stats_lost_total{datapath="ovs-system"} 3`,
+				`openvswitch_datapath_stats_misses_total{datapath="ovs-system"} 2`,
+			},
+		},
+	}
+
```
Comment (TimSimmons) [inline internal/ovsexporter/datapath_test.go:93]: This is amazing.

File: internal/ovsexporter/ovsexporter_test.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	if err != nil {
+		t.Fatalf("failed to GET data from prometheus: %v", err)
+	}
+	defer resp.Body.Close()
+
+	buf, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		t.Fatalf("failed to read server response: %v", err)
+	}
+
```
Comment (TimSimmons) [inline internal/ovsexporter/ovsexporter_test.go:51]: This is also incredible.

**Review by TimSimmons — APPROVED** [review]
This awesome and I'm totally stealing a couple of these patterns for later work. I had one question for my edification, but it's nothing blocking.

File: internal/ovsexporter/datapath.go [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+			prometheus.BuildFQName(namespace, subsystem, "info"),
+			"Metadata about a given datapath.",
+			// TODO(mdlayher): expose features?
```
Comment (si74) [inline internal/ovsexporter/datapath.go]: what features would these be?

File: internal/ovsexporter/ovsexporter.go (hunk 2 above)
Comment (si74) [reply internal/ovsexporter/ovsexporter.go:42]: Ditto this question

**Comment by si74** [comment]
Overall LGTM but please add a README.md on usage first perhaps

**Review by si74 — DISMISSED** [review]
👍 but readme s'il vous plait!

File: internal/ovsexporter/ovsexporter.go (hunk 2 above)
Comment (mdlayher) [reply internal/ovsexporter/ovsexporter.go:42]: Tim is exactly right. There are multiple generic netlink OVS families that may provide useful stats, but I'm just using the datapath one so far.

File: internal/ovsexporter/datapath.go (hunk 5 above)
Comment (mdlayher) [reply internal/ovsexporter/datapath.go]: Yeah so it looks like some netlink implementation detail stuff that I'm honestly not sure about, so my gut feeling is that we should omit them until we find some use for them.

**Comment by mdlayher** [comment]
I added a light readme previously to the top level of the repo but definitely plan to expand on it with useful PromQL queries, dashboards, etc. I'll file an issue to follow up.

File: internal/ovsexporter/datapath.go (hunk 5 above)
Comment (mdlayher) [reply internal/ovsexporter/datapath.go]: I'll probably just remove this info metric and Todo for now.

File: internal/ovsexporter/datapath_test.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+				`openvswitch_datapath_megaflow_stats_mask_hits_total{datapath="ovs-system"} 5`,
+				`openvswitch_datapath_megaflow_stats_masks_total{datapath="ovs-system"} 6`,
+				`openvswitch_datapath_stats_flows{datapath="ovs-system"} 4`,
+				`openvswitch_datapath_stats_hits_total{datapath="ovs-system"} 1`,
+				`openvswitch_datapath_stats_hits_total{datapath="ovs-test"} 99`,
+				`openvswitch_datapath_stats_lost_total{datapath="ovs-system"} 3`,
+				`openvswitch_datapath_stats_misses_total{datapath="ovs-system"} 2`,
+			},
+		},
+	}
```
Comment (mpapi) [inline internal/ovsexporter/datapath_test.go:92]: If it's straightforward to do so, what do you think about testing the case where the function to list datapaths returns an error?

**Review by mpapi — APPROVED** [review]
LGTM!

File: internal/ovsexporter/datapath_test.go (hunk 6 above)
Comment (mdlayher) [reply internal/ovsexporter/datapath_test.go:92]: Filed #3 to follow up.

**Pull request merged**